### PR TITLE
added methods to read test data from external source

### PR DESCRIPTION
**Pull request by karthikmsqa**
# Description

This PR includes the following proposed change(s):

- Methods to get the test data from the external source to the cucumber test steps. 

- [FLA-167](https://justice.gov.bc.ca/jira/browse/FLA-167)

## Type of change
- [x] Refactoring / Documentation

if your change is a breaking change, please add `breaking change` label to this PR

## How Has This Been Tested?

Tested locally
## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: tests/src/test/java/features/frontendfeatures/landingPage.feature [hunk 1]
```diff
@@ -5,13 +5,14 @@ Feature: User can navigate to the efiling landing page
...
   @frontend
   Scenario: Verify valid account guid redirects efiling form page
-    When  user enters a valid account guid "enter valid guid"
+    When  user enters a valid account guid "validGuid"
     Then eFiling frontend page is displayed and cancel button exists
+
   @frontend
   Scenario: Verify non existing account guid redirects to the efiling form page
-    When  user enters non existing account guid "123"
+    When  user enters non existing account guid "incorrectGuid"
```
Comment (sdevalapurkar-bcgov) [inline tests/src/test/java/features/frontendfeatures/landingPage.feature:13]:
```suggestion
    When  user enters non existing CSO account guid "nonExistingCSOGuid"
```

File: tests/src/test/java/features/frontendfeatures/landingPage.feature [hunk 2]
```diff
@@ -5,13 +5,14 @@ Feature: User can navigate to the efiling landing page
 
   @frontend
   Scenario: Verify valid account guid redirects efiling form page
-    When  user enters a valid account guid "enter valid guid"
+    When  user enters a valid account guid "validGuid"
```
Comment (sdevalapurkar-bcgov) [inline tests/src/test/java/features/frontendfeatures/landingPage.feature:8]:
```suggestion
    When  user enters a valid existing CSO account guid "validExistingCSOGuid"
```

File: tests/src/test/java/features/frontendfeatures/landingPage.feature [hunk 3]
```diff
@@ -5,13 +5,14 @@ Feature: User can navigate to the efiling landing page
...
+
   @frontend
   Scenario: Verify non existing account guid redirects to the efiling form page
-    When  user enters non existing account guid "123"
+    When  user enters non existing account guid "incorrectGuid"
     Then eFiling frontend page is displayed and cancel button exists
   @frontend
   Scenario: Verify existing account guid with conflicts returns error message
-    When  user enters invalid account guid "enter invalid guid"
+    When  user enters invalid account guid "invalidGuid"
```
Comment (sdevalapurkar-bcgov) [inline tests/src/test/java/features/frontendfeatures/landingPage.feature:17]:
```suggestion
    When  user enters invalid CSO account (no eFiling role) guid "invalidNoRoleGuid"
```

File: tests/src/main/java/ca/bc/gov/open/jagefilingapi/qa/frontendutils/JsonReader.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import java.util.Map;
+
+public class JsonReader {
+
+    public static Object[][] searchJsonElement(JsonArray jsonArray, int totalDataRow, int totalColumnEntry) {
```
Comment (sdevalapurkar-bcgov) [inline tests/src/main/java/ca/bc/gov/open/jagefilingapi/qa/frontendutils/JsonReader.java:14]: question: what is this doing??

File: tests/src/main/java/ca/bc/gov/open/jagefilingapi/qa/frontendutils/JsonReader.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        Object[][] matrix = new Object[totalDataRow][totalColumnEntry];
+        int i = 0;
```
Comment (akroon3r) [inline tests/src/main/java/ca/bc/gov/open/jagefilingapi/qa/frontendutils/JsonReader.java:16]: I think we can define these on one line? also I'm a sucker for descriptive counter names =P

File: tests/src/test/java/stepDefinitions/frontendstepdefinitions/LandingPageTest.java [hunk 6]
```diff
@@ -11,12 +12,18 @@
...
+import java.text.ParseException;
 
 public class LandingPageTest extends DriverClass {
 
      LandingPage landingPage;
      EfilingPage eFilingPage;
 
+    //DataProvider
+    public static Object[][] accountData() throws IOException {
+        return JsonReader.getData("src/test/java/testdatasource/account-data.json", "accountGuid", 1, 3);
```
Comment (akroon3r) [inline tests/src/test/java/stepDefinitions/frontendstepdefinitions/LandingPageTest.java:24]: what do the values 1 and 3 refer to?

File: tests/src/test/java/testdatasource/account-data.json [hunk 7]
```diff
@@ -0,0 +1,9 @@
+{
+  "accountGuid": [
+    {
+      "validGuid": "<validGuid>",
+      "invalidGuid": "<invalidGuid>",
+      "incorrectGuid": "<IncorrectGuid>"
```
Comment (akroon3r) [inline tests/src/test/java/testdatasource/account-data.json:6]: following the values above, I believe the first char should be lower case here :)

File: tests/README.md [hunk 8]
```diff
@@ -21,7 +21,8 @@ tests
 
 ## Running the maven Tests without cucumber tags
 
-Add values to the keys in config.properties file
+1. Add values to the keys in config.properties file
```
Comment (akroon3r) [inline tests/README.md:24]: The markup suggestion here from code climate. Not sure if you've seen or dealt with the warnings it places inline on the diff yet. Click on the yellow bar to expand the message.

File: tests/src/test/java/features/frontendfeatures/landingPage.feature (hunk 1 above)
Comment (karthikmsqa) [reply tests/src/test/java/features/frontendfeatures/landingPage.feature:13]: Done. Thanks!

File: tests/src/test/java/features/frontendfeatures/landingPage.feature (hunk 3 above)
Comment (karthikmsqa) [reply tests/src/test/java/features/frontendfeatures/landingPage.feature:17]: Done. Thanks!

File: tests/src/main/java/ca/bc/gov/open/jagefilingapi/qa/frontendutils/JsonReader.java (hunk 4 above)
Comment (karthikmsqa) [reply tests/src/main/java/ca/bc/gov/open/jagefilingapi/qa/frontendutils/JsonReader.java:14]: For sending multiple test datasets in the future, treating datasets in the array as rows and each key/value pairs as columns like excel and iterating through the json array to parse. Hope that makes sense.

File: tests/src/test/java/stepDefinitions/frontendstepdefinitions/LandingPageTest.java (hunk 6 above)
Comment (karthikmsqa) [reply tests/src/test/java/stepDefinitions/frontendstepdefinitions/LandingPageTest.java:24]: 1 refers to total no of data set in the 'accountGuid' array, in the current file its 1.
3 refers to total no of rows(key/value) pairs in the particular dataset in the current file its 3. 
Hope that makes sense

File: tests/src/main/java/ca/bc/gov/open/jagefilingapi/qa/frontendutils/JsonReader.java (hunk 4 above)
Comment (sdevalapurkar-bcgov) [reply tests/src/main/java/ca/bc/gov/open/jagefilingapi/qa/frontendutils/JsonReader.java:14]: yeah it does, just wondering if/how we can simplify this

File: tests/src/test/java/testdatasource/account-data.json (hunk 7 above)
Comment (alexjoybc) [reply tests/src/test/java/testdatasource/account-data.json:6]: you can use an object like this

```java
public class AccountDataSet {

    @JsonProperty("AccountData ")
    private Map<String, String> AccountData = null;

    // getters and setters

}
```

then in the code

```java 

String json = "{ \"accountData\" : [(\"validGuid\":\"your values\" }]";
AccountDataSet accountDataSet = objectMapper.readValue(json, AccountDataSet .class); 
String validGuid = accountDataSet .getAccountData.get("validGuid");

```

File: tests/src/test/java/features/frontendfeatures/landingPage.feature (hunk 2 above)
Comment (karthikmsqa) [reply tests/src/test/java/features/frontendfeatures/landingPage.feature]: done. Thanks

File: tests/src/main/java/ca/bc/gov/open/jagefilingapi/qa/frontendutils/JsonReader.java (hunk 4 above)
Comment (karthikmsqa) [reply tests/src/main/java/ca/bc/gov/open/jagefilingapi/qa/frontendutils/JsonReader.java]: 2D array was approach was to work with the @dataprovider annotaion. Refactored now. Thank you.

File: tests/src/main/java/ca/bc/gov/open/jagefilingapi/qa/frontendutils/JsonReader.java (hunk 5 above)
Comment (karthikmsqa) [reply tests/src/main/java/ca/bc/gov/open/jagefilingapi/qa/frontendutils/JsonReader.java]: Refactored. Thanks.

File: tests/src/test/java/stepDefinitions/frontendstepdefinitions/LandingPageTest.java (hunk 6 above)
Comment (karthikmsqa) [reply tests/src/test/java/stepDefinitions/frontendstepdefinitions/LandingPageTest.java]: Refactored code uses POJO and jackson bind. Thank you.

File: tests/src/test/java/testdatasource/account-data.json (hunk 7 above)
Comment (karthikmsqa) [reply tests/src/test/java/testdatasource/account-data.json]: Updated. Thanks.

File: tests/README.md (hunk 8 above)
Comment (karthikmsqa) [reply tests/README.md]: done. Thanks.

File: tests/README.md [hunk 9]
```diff
@@ -21,7 +21,8 @@ tests
...
+2. Add test data values to the json files in 'test/java/testdatasource' package
```
[the remark this answers is not in the document]
Comment (karthikmsqa) [reply tests/README.md]: updated

File: tests/pom.xml [hunk 10]
```diff
@@ -100,6 +100,11 @@
       <artifactId>cucumber-java</artifactId>
       <version>${cucumber.version}</version>
     </dependency>
+    <dependency>
+      <groupId>io.cucumber</groupId>
+      <artifactId>cucumber-core</artifactId>
+      <version>${cucumber.version}</version>
```
[the remark this answers is not in the document]
Comment (karthikmsqa) [reply tests/pom.xml]: updated to use  {[groupId.version}. Thanks.

File: tests/pom.xml [hunk 11]
```diff
@@ -118,18 +123,6 @@
       <version>1.0.0</version>
       <scope>test</scope>
     </dependency>
-    <dependency>
-      <groupId>org.awaitility</groupId>
-      <artifactId>awaitility</artifactId>
-      <version>3.0.0</version>
```
[the remark this answers is not in the document]
Comment (karthikmsqa) [reply tests/pom.xml:124]: This dependency is removed.

File: tests/src/main/java/ca/bc/gov/open/jagefilingapi/qa/frontendutils/JsonDataReader.java [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+        throw new IllegalStateException("Json Reader Utility class");
+    }
+
+    public static CsoAccountGuid getCsoAccountGuid() throws IOException {
+
+        ObjectMapper mapper = new ObjectMapper();
+        InputStream inputStream = new FileInputStream("src/test/java/testdatasource/account-data.json");
+        TypeReference<List<CsoAccountGuid>> typeReference= new TypeReference<List<CsoAccountGuid>>() {};
+        List<CsoAccountGuid> csoAccountGuids = mapper.readValue(inputStream, typeReference);
+        CsoAccountGuid accounts = null;
```
Comment (sdevalapurkar-bcgov) [inline tests/src/main/java/ca/bc/gov/open/jagefilingapi/qa/frontendutils/JsonDataReader.java:22]: isnt this a single account? so should it be called `account`?

File: tests/src/main/java/ca/bc/gov/open/jagefilingapi/qa/testdatatypes/CsoAccountGuid.java [hunk 13]
```diff
@@ -0,0 +1,32 @@
+package ca.bc.gov.open.jagefilingapi.qa.testdatatypes;
+
+public class CsoAccountGuid {
+
+    private String validExistingCSOGuid;
+    private String invalidGuid;
```
Comment (sdevalapurkar-bcgov) [inline tests/src/main/java/ca/bc/gov/open/jagefilingapi/qa/testdatatypes/CsoAccountGuid.java:6]: I still think it is better to name this as:
```
invalidNoFilingRoleGuid
```
or something like that so it is clear

File: tests/src/test/java/testdatasource/account-data.json [hunk 14]
```diff
@@ -0,0 +1,7 @@
+[
+  {
+    "validExistingCSOGuid": "<validExistingCSOGuid>",
+    "invalidGuid": "<invalidGuid>",
```
Comment (sdevalapurkar-bcgov) [inline tests/src/test/java/testdatasource/account-data.json:4]: same here and in other locations

File: tests/src/main/java/ca/bc/gov/open/jagefilingapi/qa/frontendutils/JsonDataReader.java (hunk 12 above)
Comment (karthikmsqa) [reply tests/src/main/java/ca/bc/gov/open/jagefilingapi/qa/frontendutils/JsonDataReader.java:22]: Updated. Thanks!

File: tests/src/main/java/ca/bc/gov/open/jagefilingapi/qa/testdatatypes/CsoAccountGuid.java (hunk 13 above)
Comment (karthikmsqa) [reply tests/src/main/java/ca/bc/gov/open/jagefilingapi/qa/testdatatypes/CsoAccountGuid.java:6]: Updated. Thanks!

File: tests/src/test/java/testdatasource/account-data.json (hunk 14 above)
Comment (karthikmsqa) [reply tests/src/test/java/testdatasource/account-data.json:4]: done.

**Pull request merged**